pkg/operator/apiserver/controller/apiservice/apiservice_controller.go

```diff
@@ -0,0 +1,234 @@
+package apiservice
+
+import (
+	"context"
+	"fmt"
+	"sort"
+	"strings"
+	"time"
+
+	operatorsv1 "github.com/openshift/api/operator/v1"
+	operatorv1 "github.com/openshift/api/operator/v1"
+	"github.com/openshift/library-go/pkg/operator/events"
+	"github.com/openshift/library-go/pkg/operator/resource/resourceapply"
+	"github.com/openshift/library-go/pkg/operator/status"
+	"github.com/openshift/library-go/pkg/operator/v1helpers"
+	"k8s.io/apimachinery/pkg/util/errors"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	kubeinformers "k8s.io/client-go/informers"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog"
+	apiregistrationv1 "k8s.io/kube-aggregator/pkg/apis/apiregistration/v1"
+	apiregistrationv1client "k8s.io/kube-aggregator/pkg/client/clientset_generated/clientset/typed/apiregistration/v1"
+	apiregistrationinformers "k8s.io/kube-aggregator/pkg/client/informers/externalversions"
+)
+
+const (
+	workQueueKey = "key"
+)
+
+type APIServiceController struct {
+	name        string
+	apiServices []*apiregistrationv1.APIService
+	// precondition must return true before the apiservices will be created
+	precondition wait.ConditionFunc
+
+	versionRecorder         status.VersionGetter
+	operatorClient          v1helpers.OperatorClient
+	kubeClient              kubernetes.Interface
+	apiregistrationv1Client apiregistrationv1client.ApiregistrationV1Interface
+	eventRecorder           events.Recorder
+
+	// queue only ever has one item, but it has nice error handling backoff/retry semantics
+	queue workqueue.RateLimitingInterface
+}
+
+func NewAPIServiceController(
```

Comment:
this has slightly changed in https://github.com/openshift/cluster-openshift-apiserver-operator/pull/294

worth picking as it made it more generic.